makelove/macos.py

```diff
@@ -0,0 +1,245 @@
+
+    sizetotypes = {
+        16: [b'icp4'],  # 16x16   std only  (no 8x8@2x)
+        32: [b'icp5', b'ic11'],  # 32x32   std -AND- 16x16@2x   high
+        64: [b'icp6', b'ic12'],  # 64x64   std -AND- 32x32@2x   high
+        128: [b'ic07'],  # 128x128 std only  (no 64x64@2x)
+        256: [b'ic08', b'ic13'],  # 256x256 std -AND- 128x128@2x high
+        512: [b'ic09', b'ic14'],  # 512x512 std -AND- 256x256@2x high
+        1024: [b'ic10']  # 1024x1024 (10.7) = 512x512@2x high (10.8)
+    }
+
+    imagedatas = []
+    for size_px, icontypes in sizetotypes.items():
+        img = icon_image.resize((size_px, size_px), Image.LANCZOS)
+        with io.BytesIO() as img_data_f:
+            img.save(img_data_f, 'png')
+            for icontype in icontypes:
+                imagedatas.append([icontype, img_data_f.getvalue()])
+
+    # 1) HEADER: 4-byte "magic" + 4-byte filesize (including header itself)
+
+    filelen = 8 + sum(
+        len(imagedata) + 8
+        for (_, imagedata) in sorted(imagedatas)
+    )
+    iconfile.write(b'icns')
+    iconfile.write(struct.pack('>I', filelen))
+
+    # 2) IMAGE TYPE+LENGTH+BYTES: packed into rest of icon file sequentially
+
+    for icontype, imagedata in imagedatas:
+        # data length includes type and length fields (4+4)
+        iconfile.write(icontype)  # 4 byte type
+        iconfile.write(struct.pack('>I', 8 + len(imagedata)))  # 4-byte length
+        iconfile.write(imagedata)  # and the image
+
+
+def get_game_icon_content(config):
+    # Mac icons are not supposed to take up the full image area and generally
+    # have shadows, etc - allow users to provide a different design but fall
+    # back on the generic icon_file setting
+    icon_file = config.get("macos", {}).get('icon_file')
+    if icon_file is None:
+        icon_file = config.get("icon_file", None)
+    elif not os.path.isfile(icon_file):
+        sys.exit(f"Couldn't find macOS icon_file at {icon_file}")
+
+    if icon_file is None:
+        icon_file = config.get("icon_file", None)
+    elif not os.path.isfile(icon_file):
+        sys.exit(f"Couldn't find icon_file at {icon_file}")
+
+    if not icon_file:
+        return False
+
+    with io.BytesIO() as icns_f, open(icon_file, 'rb') as icon_img_f:
+        icon_key = f"{config['name']}.app/Contents/Resources/icon-{config['name']}.icns"
+        if icon_file.lower().endswith('.png'):
+            make_icns(icns_f, icon_img_f)
+            return icns_f.getvalue()
+        else:
+            return icon_img_f.read()
+
+
+def get_info_plist_content(config, version):
+    metadata = {}
+    if "macos" in config and "app_metadata" in config["macos"]:
+        metadata = config["macos"]["app_metadata"]
+
+    metadata.setdefault('FileVersion', version or config['love_version'])
+    metadata.setdefault('ProductName', config['name'])
+    metadata.setdefault('LegalCopyright', "© 2006-2020 LÖVE Development Team")
+    metadata.setdefault('BundleIdentifier', f"tld.{config['name']}")
+
```

Comment:
If you can't reasonably process every possible value put into `app_metadata`, I think you should log something if you encounter an unknown key.

Reply:
I've updated this to take the actual keys used in the plist and now allow users to specify arbitrary plist keys